### DISCUSSION TITLE
101. MySQL(InnoDB)에서 인덱스는 어디에 저장되나요?

### DISCUSSION THREAD

**Discussion by jongmee** [jewelry-box/cs-interview#105] [General]
1. MySQL에서 인덱스는 어디에 저장되나요?
2. 인덱스의 저장 단위는 무엇인가요?
3. 페이지 크기와 인덱스의 성능 관계가 어떻게 될까요?

키워드: 인덱스 저장 구조, 저장 단위, 페이지 크기와 인덱스 성능의 상관관계

**Comment by cutehumanS2** [comment]
### InnoDB에서 인덱스는 어디에 저장되나요?

InnoDB의 인덱스는 **디스크(테이블스페이스 파일)에** **저장**되며, **메모리(버퍼 풀)에 캐싱**되어 읽기 및 쓰기 성능을 높입니다.

- InnoDB는 데이터를 저장하는 테이블스페이스에 인덱스를 함께 저장합니다.
    - 클러스터형 인덱스는 데이터 자체와 함께 물리적으로 정렬된 상태로 저장됩니다.
    - 세컨더리 인덱스(보조 인덱스)는 인덱스 키와 함께 해당 클러스터형 인덱스의 값(PK)을 가리키는 정보도 저장됩니다.
    - 이 정보는 기본적으로 테이블의 .ibd 파일 또는 ibdata1 파일(공유  테이블스페이스 모드) 내에 저장됩니다.
- InnoDB는 디스크에 저장된 인덱스를 메모리에 캐싱하여 성능을 최적화합니다.
    - 버퍼 풀이라는 메모리 영역에 자주 사용되는 인덱스 페이지와 데이터 페이지를 캐싱합니다. 이를 통해 디스트 I/O를 최소화하고 쿼리 성능을 크게 향상시킵니다.

<br>

[참고한 자료]
- https://dev.mysql.com/blog-archive/innodb-tablespace-space-management/
- https://dev.mysql.com/doc/refman/8.4/en/innodb-file-per-table-tablespaces.html
- 책 <Real MySQL 8.0 (1권)> - 4.2.7 InnoDB 버퍼 풀

**Comment by cutehumanS2** [comment]
### InnoDB에서 인덱스의 저장 단위는 무엇인가요?
- InnoDB에서 인덱스는 페이지 단위로 디스크에 저장되며, 기본 16KB입니다.
- B+트리 구조에서 각 노드가 하나의 페이지로 관리됩니다.
- 자주 사용되는 페이지는 버퍼 풀에 캐싱되어 디스크 I/O를 줄이고 성능을 최적화합니다.

<br>

[참고한 자료]
책 <Real MySQL 8.0 (1권)> - 8.3.3.1 인덱스 키 값의 크기

**Comment by cutehumanS2** [comment]
### InnoDB에서 페이지 크기가 인덱스 성능에 어떤 영향을 미치나요?

- **페이지 크기가 크면**
    - 한 번의 I/O로 더 많은 데이터와 인덱스 엔트리를 읽을 수 있어, 순차적 읽기나 대량 데이터 처리에서 유리합니다.
    - 트리의 깊이가 낮아져 탐색 단계가 줄어들어 검색 속도가 빨라집니다.
    - 하지만 자주 변경되는 데이터에서는 페이지 단위 잠금 충돌 가능성이 높아지고, 불필요한 데이터를 많이 읽어와 랜덤 읽기 효율이 떨어질 수 있습니다.
- **페이지 크기가 작으면**
    - 데이터 변경 시 영향을 받는 범위가 적어져 동시성 처리에서 유리합니다.
    - 적은 데이터만 읽어와, 랜덤 읽기 성능이 개선됩니다.
    - 하지만 I/O 작업이 빈번해져 디스크 접근 비용이 증가하고, 트리 깊이가 깊어져 탐색 속도가 느려질 수 있습니다.

따라서 시스템의 특성(읽기/쓰기 패턴)에 따라 적절한 페이지 크기를 선택하는 게 중요합니다.

<br>

[참고한 자료]
- 책 <Real MySQL 8.0 (1권)> - 8.3.3.1 인덱스 키 값의 크기
- 책 <Real MySQL 8.0 (1권)> - 8.3.3.2 B-Tree 깊이
